File: ios/AudioCue/AudioCueModule.swift
import AVFoundation

/**
 * AudioCue TurboModule — iOS implementation
 *
 * Synthesizes a short sine tone directly into a PCM buffer and plays it via AVAudioEngine /
 * AVAudioPlayerNode - no bundled audio files, no Xcode Resources entries, nothing read from disk.
 * A new/different tone is just a different (frequency, duration) call from JS; nothing here or in
 * the Xcode project ever needs to change for it.
 *
 * Deliberately uses only the AVAudioSession playback surface (.playback category, .mixWithOthers
 * option) and AVAudioEngine's output-side nodes (player -> mixer -> output) — never touches
 * AVAudioSession's record-permission APIs or AVAudioEngine's input node, so no
 * NSMicrophoneUsageDescription entry is required. Replaces a prior dependency on
 * react-native-audio-api, whose native module linked AVAudioApplication/AVAudioSession
 * record-permission symbols regardless of whether the app called them, which App Store review
 * (ITMS-90683) flags on binary presence alone.
 */
@objc(AudioCue)
class AudioCueModule: NSObject {

    // 5ms gain ramp-down at the end of every tone, to avoid an audible click when the waveform is
    // cut off mid-cycle rather than at a zero crossing.
    private static let rampSeconds = 0.005

    private let engine = AVAudioEngine()
    private let playerNode = AVAudioPlayerNode()

    private var sessionConfigured = false
    private var connectedFormat: AVAudioFormat?

    @objc(play:durationMs:)
    func play(_ frequencyHz: NSNumber, durationMs: NSNumber) {
        configureSessionIfNeeded()

        guard let format = startEngineIfNeeded() else {
            return
        }

        let sampleRate = format.sampleRate
        let durationSec = durationMs.doubleValue / 1000.0
        let frameCount = AVAudioFrameCount(max(0, durationSec) * sampleRate)

        guard frameCount > 0,
              let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: frameCount) else {
            return
        }
        buffer.frameLength = frameCount

        let frequency = frequencyHz.doubleValue
        let rampFrames = min(AVAudioFrameCount(AudioCueModule.rampSeconds * sampleRate), frameCount / 2)
        let channelData = buffer.floatChannelData![0]

        for frame in 0..<Int(frameCount) {
            let time = Double(frame) / sampleRate
            let framesFromEnd = Int(frameCount) - frame
            let amplitude = (rampFrames > 0 && framesFromEnd < Int(rampFrames))
                ? Double(framesFromEnd) / Double(rampFrames)
                : 1.0
            channelData[frame] = Float(sin(2.0 * Double.pi * frequency * time) * amplitude)
        }

        playerNode.scheduleBuffer(buffer, completionHandler: nil)
        if !playerNode.isPlaying {
            playerNode.play()
        }
    }

    // Configures the session so tones mix with whatever media/music is already playing instead of
    // being silenced by the mute switch. Best-effort - a failure here must never block playback.
    private func configureSessionIfNeeded() {
        guard !sessionConfigured else { return }
        sessionConfigured = true

        do {
            let session = AVAudioSession.sharedInstance()
            try session.setCategory(.playback, options: [.mixWithOthers])
            try session.setActive(true)
        } catch {
            // ignore - tones still attempt to play without the preferred session configuration
        }
    }

    // Attaches/connects the player node and starts the engine exactly once, using the hardware's
    // actual output sample rate. Returns the mono format every scheduled buffer must be built
    // against - reused on every call rather than recomputed, since AVAudioEngine only allows one
    // active connection format per node pair.
    private func startEngineIfNeeded() -> AVAudioFormat? {
        if let format = connectedFormat {
            if !engine.isRunning {
                try? engine.start()
            }
            return format
        }

        let sampleRate = engine.outputNode.outputFormat(forBus: 0).sampleRate
        guard sampleRate > 0,
              let format = AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1) else {
            return nil
        }

        engine.attach(playerNode)
        engine.connect(playerNode, to: engine.mainMixerNode, format: format)

        do {
            try engine.start()
        } catch {
            return nil
        }

        connectedFormat = format
        return format
    }

    @objc static func requiresMainQueueSetup() -> Bool {
        return false
    }
}

File: android/app/src/main/java/com/incyclist/app/AudioCueModule.kt
package com.incyclist.app

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.module.annotations.ReactModule
import kotlin.math.PI
import kotlin.math.min
import kotlin.math.sin

/**
 * AudioCue TurboModule — Android implementation.
 *
 * Synthesizes a short sine tone directly into a PCM buffer and plays it via AudioTrack - no
 * bundled audio files, no res/raw or assets/ entries, nothing read from disk. A new/different tone
 * is just a different (frequencyHz, durationMs) call from JS; nothing here or in the Android
 * project ever needs to change for it.
 *
 * Playback-only — never requests RECORD_AUDIO or touches any microphone API, unlike the prior
 * react-native-audio-api dependency this replaces.
 */
@ReactModule(name = AudioCueModule.NAME)
class AudioCueModule(reactContext: ReactApplicationContext) :
    NativeAudioCueSpec(reactContext) {

    override fun getName(): String = NAME

    private val sampleRate = 44100

    // 5ms gain ramp-down at the end of every tone, to avoid an audible click when the waveform is
    // cut off mid-cycle rather than at a zero crossing.
    private val rampSeconds = 0.005

    override fun play(frequencyHz: Double, durationMs: Double) {
        try {
            val frameCount = ((durationMs / 1000.0) * sampleRate).toInt()
            if (frameCount <= 0) return

            val samples = renderSamples(frequencyHz, frameCount)

            val audioTrack = AudioTrack.Builder()
                .setAudioAttributes(
                    AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_MEDIA)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                        .build()
                )
                .setAudioFormat(
                    AudioFormat.Builder()
                        .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                        .setSampleRate(sampleRate)
                        .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                        .build()
                )
                .setBufferSizeInBytes(samples.size * 2)
                .setTransferMode(AudioTrack.MODE_STATIC)
                .build()

            audioTrack.write(samples, 0, samples.size)

            // Short one-shot tone: release the track once playback reaches the end rather than
            // leaking it - there is no other natural "done" signal for a MODE_STATIC track.
            audioTrack.setNotificationMarkerPosition(frameCount)
            audioTrack.setPlaybackPositionUpdateListener(object : AudioTrack.OnPlaybackPositionUpdateListener {
                override fun onMarkerReached(track: AudioTrack) {
                    track.release()
                }
                override fun onPeriodicNotification(track: AudioTrack) {
                    // Required by the interface, but only onMarkerReached is used - no periodic
                    // notification period is set, so this is never actually called.
                }
            })

            audioTrack.play()
        } catch (e: Exception) {
            // never let a tone-playback failure take down an in-progress ride
        }
    }

    private fun renderSamples(frequencyHz: Double, frameCount: Int): ShortArray {
        val samples = ShortArray(frameCount)
        val rampFrames = min((rampSeconds * sampleRate).toInt(), frameCount / 2)

        for (i in 0 until frameCount) {
            val time = i.toDouble() / sampleRate
            val framesFromEnd = frameCount - i
            val amplitude = if (rampFrames > 0 && framesFromEnd < rampFrames) {
                framesFromEnd.toDouble() / rampFrames
            } else {
                1.0
            }
            samples[i] = (sin(2.0 * PI * frequencyHz * time) * amplitude * Short.MAX_VALUE).toInt().toShort()
        }

        return samples
    }

    companion object {
        const val NAME = "AudioCue"
    }
}
